frontend/util.lua

```diff
@@ -1635,4 +1635,15 @@ function util.round_decimal(num, points)
     return math.floor(num * op) / op
 end
 
+-- Check if a certain command is executable (e.g. ntpd or ntp for Time Sync plugin)
+function util.which(command)
+    local path = os.getenv("PATH") or ""
+    for p in path:gmatch("([^:]+)") do
```

Comment:
This is technically buggy in that in a `PATH` like `/usr/bin:` the empty component is not honored (instead of looking into the current directory).

But anyway, we could add a optional `path` argument, and then use it the terminal plugin in place of `isExecutable`:

```lua
local function which(command, path)
    path = path or os.getenv("PATH") or ""
    for p in path:gmatch("([^:]+)") do
        p = p .. "/" .. command
        if lfs.attributes(p) and os.execute("test -x " .. p) == 0 then
            return p
        end
    end
end

[…]

function Terminal:getDefaultShellExecutable()
    if self.default_shell_executable then return self.default_shell_executable end

    local shell = {
        "bash",
        "ash",
        "sh",
        "zsh",  -- RPROMPTs aren't really handled well, so we deprioritize it a bit
        "dash",
        "hush",
        "ksh",
        "mksh",
    }
    local env_shell = os.getenv("SHELL")
    if env_shell then
        table.insert(shell, 1, env_shell)
    end

    local path = ".:" .. (os.getenv("PATH") or "")
    for dummy, file in ipairs(shell) do
        if which(file, path) then
            self.default_shell_executable = file
            break
        end
    end
    logger.dbg("Terminal: default shell is", self.default_shell_executable)

    return self.default_shell_executable
end
```

Reply:
Should I also alter `util.which` function to add `path argument, and then refer to it in the terminal emulator plugin?

Reply:
> Should I also alter `util.which` function to add `path` argument,

Yes.

> and then refer to it in the terminal emulator plugin?

I don't think it necessary to refer to the terminal plugin (or the timesync plugin, for that matter).

Reply:
You mean the comment in util? I can remove it.
By "refer to it" I meant using `util.which` function call

Reply:
And for time sync, is it fine to have second argument missing? Like `local ntpd = util.which('ntpd')`, etc

Reply:
Better:

```lua
function util.isExecutable(file)
    local attrs = lfs.attributes(file)
    return attrs and attrs.mode == "file" and os.execute("test -x " .. file) == 0
end

function util.which(command, path)
    path = path or os.getenv("PATH") or ""
    for p in path:gmatch("([^:]+)") do
        if util.isExecutable(p .. "/" .. command) then
            return p
        end
    end
end
```

Then in the terminal plugin just fix `Terminal:isExecutable`:
```diff
 plugins/terminal.koplugin/main.lua | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git i/plugins/terminal.koplugin/main.lua w/plugins/terminal.koplugin/main.lua
index 347f53050..3a0eaba65 100644
--- i/plugins/terminal.koplugin/main.lua
+++ w/plugins/terminal.koplugin/main.lua
@@ -6,6 +6,7 @@ This plugin provides a terminal emulator (VT52 (+some ANSI and some VT100))
 
 local Device = require("device")
 local logger = require("logger")
+local util = require("util")
 local ffi = require("ffi")
 local C = ffi.C
 require("ffi/posix_h")
@@ -94,7 +95,7 @@ local Terminal = WidgetContainer:extend{
 
 function Terminal:isExecutable(file)
     -- check if file is an executable or a command in PATH
-    return os.execute(string.format("test -x %s || command -v %s", file, file)) == 0
+    return util.isExecutable(file) or util.which(file)
 end
 
 -- Try SHELL environment variable and some standard shells
```

Reply:
Ah, sorry, I didn't refresh the page before posting the above reply.
One second...

Reply:
Great success! crash.log is again reporting `ash`:
```
11/21/24-00:25:25 DEBUG Terminal: default shell is ash 
11/21/24-00:25:25 DEBUG RD loaded plugin terminal at plugins/terminal.koplugin 
```
and just `ash` works well now!
Thank you!
I'll commit these changes